refactor: remove all hardcoded values from clothes suggestion use case

norhanadel367 opened the pull request.

norhanadel367 committed:
refactor: improve ClothingSuggestionUseCase

abdosharaf9 commented on src/main/kotlin/logic/model/Clothes.kt:
```diff
@@ -0,0 +1,6 @@
+package logic.model
+
+data class Clothes(
+    val nameOfClothes : String,
```
It can be `name` only, because it can be understood from the class context.

Shehab-m replied:
Agree with sharaf it could be name or clotheName

abdosharaf9 commented on src/main/kotlin/logic/uitls/DataSource.kt:
```diff
@@ -0,0 +1,21 @@
+package logic.uitls
```
1. There is a typo in the word `utils`.
2. About the `DataSource`:
    - I think the name is misleading because it only holds the clothes list, not the whole data source.
    - IDK if it should stay `object`, if it can be changed to a function/class, or at least only the list with itself in the file without any container.
3. I guess this should be in the data package, not the logic.
4. Note that you used hardcoded values for the tags.

norhanadel367 replied:
if move it to data ? logic depend on data ? i think in logic . and rename it only ?

abdosharaf9 replied:
@Shehab-m 
What do u think about this?

Shehab-m replied:
It should be in data as it is a data u fetch for the user can be replaced in the future with api or anything and it doesn't mean we will make the domain depend on the data it will be abstraction as is

abdosharaf9 commented on src/main/kotlin/logic/usecase/ClothingSuggestionUseCase.kt:
```diff
@@ -10,48 +11,31 @@ class ClothingSuggestionUseCase {
-                    listOf("Light Dress 👗")
-                }
-            }
+            temperature < WeatherConstants.FREEZING_TEMP ->
```
Your `when` statement should only decide the values you need to filter the clothes list with, and then filter the list according to the values decision. Also, improve all the hardcoded string values using constants.

Shehab-m replied:
Agree plus make it a data class that contains the clothes data not just plain lists

Shehab-m reviewed (changes requested):
Review the comments

norhanadel367 committed:
refactor: move SuggestionClothesDataSource to data package, improve ClothingSuggestionUseCase

norhanadel367 committed:
refactor: edite unit test based changed in use case ClothingSuggestionUseCaseTest

norhanadel367 committed:
Merge remote-tracking branch 'origin/base-structure' into feature/clothing-suggestion

# Conflicts:
#	src/test/kotlin/logic/usecase/ClothingSuggestionUseCaseTest.kt

abdosharaf9 commented on src/main/kotlin/logic/repository/ClothingSuggestionRepository.kt:
```diff
@@ -0,0 +1,7 @@
+import logic.model.Clothes
+
+interface ClothingSuggestionRepository {
+    fun getAllClothes(): List<Clothes>
```
I guess if you want to create a repository, you should have a function that returns the clothes filtered instead of filtering them yourself.

abdosharaf9 commented on src/main/kotlin/data/datasource/SuggestionClothesDataSource.kt:
```diff
@@ -0,0 +1,21 @@
+
+import logic.model.Clothes
+
+object SuggestionClothesDataSource {
```
For now, this is not wrong and will work, but in the future, if we have more logic in this place, such as fetching the data from an API, it must be a class. IDK if you should change it from now or not.

abdosharaf9 commented on src/main/kotlin/data/datasource/SuggestionClothesDataSource.kt:
```diff
@@ -0,0 +1,21 @@
+
+object SuggestionClothesDataSource {
+    val clothesSuggestion: List<Clothes> = listOf(
+        Clothes(name = "Coat 🧥", tags = listOf("freezing_temp", "morning", "night")),
```
Remember to use the tags you created.

norhanadel367 committed:
refactor: using kotlinx.datetime not  java.time.LocalTime . inject repo in ClothingSuggestionUseCase

norhanadel367 committed:
refactor : add fun filterClothes in repo to filter clothes in ClothingSuggestionUseCase

norhanadel367 committed:
refactor :edite fun suggestClothes , move filteration getTimeOfDay ,
getWeatherTag in WeathersUtils in logic .  inject clothingSuggestionRepository in use case.

norhanadel367 committed:
refactor : convert hardcode eith constants , add fun filterClothes

norhanadel367 committed:
refactor: make SuggestionClothes nullable value

norhanadel367 committed:
refactor: edit unit test to make weathersUtils mockk,and mockk weatherRepository ,edite test case based it

norhanadel367 committed:
Merge remote-tracking branch 'origin/base-structure' into feature/clothing-suggestion

# Conflicts:
#	src/main/kotlin/data/repository/WeatherRepositoryImpl.kt
#	src/main/kotlin/logic/usecase/ClothingSuggestionUseCase.kt
#	src/main/kotlin/logic/utils/Exceptions.kt

abdosharaf9 commented on src/main/kotlin/logic/utils/WeathersUtils.kt:
```diff
@@ -0,0 +1,25 @@
+import logic.usecase.WeatherConstants
+
+object WeathersUtils {
+    fun getTimeOfDay(currentWeather: CurrentWeather): String {
```
This can be renamed to something like `getTimeOfDayTag` for better readability.

abdosharaf9 replied:
The same thing for the whole object as an argument.

norhanadel367 replied:
The same thing for the whole object as an argument. but i think it is readable

abdosharaf9 commented on src/main/kotlin/logic/utils/WeathersUtils.kt:
```diff
@@ -0,0 +1,25 @@
+        }
+    }
+
+    fun getWeatherTag(currentWeather: CurrentWeather): String {
```
The name is misleading, because it doesn't tell that it returns the temperature tag explicitly. Also, I guess you just need the temperature as an argument not the whole object.

abdosharaf9 commented on src/main/kotlin/logic/utils/WeathersUtils.kt:
```diff
@@ -0,0 +1,25 @@
+import logic.model.CurrentWeather
+import logic.usecase.WeatherConstants
+
+object WeathersUtils {
```
I guess this needs to be a class, and be injected into the use case.

norhanadel367 replied:
two  function is only used once .so it's better to keep it as an object in the utils file instead of making it a class and injecting it.

abdosharaf9 commented on src/main/kotlin/logic/repository/WeatherRepository.kt:
```diff
@@ -6,4 +6,5 @@ import logic.model.CurrentWeather
 interface WeatherRepository {
     suspend fun getCityLocationByName(cityName: String): CityLocation
     suspend fun getWeatherFromRemote(cityLocation: CityLocation): CurrentWeather
+    fun filterClothes( weatherTag:List<String>):List<String>?
```
Why this function returns a nullable list?

norhanadel367 replied:
i think in future if used api  , if  not  match tage is return null.

abdosharaf9 commented on src/main/kotlin/logic/model/SuggestedClothes.kt:
```diff
@@ -0,0 +1,8 @@
+package logic.model
+
+data class SuggestedClothes(
+    val hour : Int? = null,
```
Remember to rename these attributes as well.

abdosharaf9 commented on src/main/kotlin/data/datasource/SuggestionClothesDataSource.kt:
```diff
@@ -0,0 +1,31 @@
+        Clothes(name = "Shorts 🩳", tags = listOf(WeatherConstants.HOT_TAG, WeatherConstants.WARM_TAG, WeatherConstants.MORNING))
+    )
+    fun filterClothes(weatherTag : List<String>) : List<String>? {
+        val matchedClothes = clothesSuggestion
```
You don't have to save the list in a variable, just do all your functions using Functional Programming on the original list.

abdosharaf9 commented on src/main/kotlin/data/datasource/SuggestionClothesDataSource.kt:
```diff
@@ -0,0 +1,31 @@
+        Clothes(name = "T-Shirt 👕", tags = listOf(WeatherConstants.HOT_TAG, WeatherConstants.WARM_TAG, WeatherConstants.MORNING, WeatherConstants.NIGHT)),
+        Clothes(name = "Shorts 🩳", tags = listOf(WeatherConstants.HOT_TAG, WeatherConstants.WARM_TAG, WeatherConstants.MORNING))
+    )
+    fun filterClothes(weatherTag : List<String>) : List<String>? {
```
Why the nullable list here?

norhanadel367 replied:
i think in future if used api , if not get list oof clothes will return null.

norhanadel367 committed:
refactor: refactor some naming

Shehab-m reviewed (changes requested):
A lot of the comments still didn't get refactored plz refactor @norhanadel367
